### PR TITLE
[JAVA][Client] New object instead of null for empty POST request

### PR DESCRIPTION
**Pull request by bmordue**
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [x] Filed the PR against the correct branch: Default: `master`.
- [x] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

It's safer to initialize the request body to an empty Object instead of null if there are no body parameters. In particular, this allows Jackson to successfully serialize a POST request with no body parameters when the Content-Type header is set to "application/json, which is set by default.

See https://github.com/swagger-api/swagger-codegen/pull/7551 for more context -- bringing this over to `openapi-generator`

cc @bbdouglas @JFCote @sreeshas @jfiala @lukoyanov @cbornet @jeff9finger

### CODE REVIEW

**Comment by jmini** [comment]
Because the "ci/circleci" if failing for some download failed errors:

```
sbt.ResolveException: download failed: joda-time#joda-time;2.2!joda-time.jar
download failed: org.scala-lang.modules#scala-xml_2.11;1.0.5!scala-xml_2.11.jar(bundle)
```

I have tried to run the java sample locally. This is not working for me:

```
mvn clean verify -f samples/client/petstore/java/rest-assured/pom.xml
```

I get these compile errors:

```
[ERROR] samples/client/petstore/java/rest-assured/src/main/java/org/openapitools/client/api/PetApi.java:[606,37] variable _FORM is already defined in class org.openapitools.client.api.PetApi.UpdatePetWithFormOper
[ERROR] samples/client/petstore/java/rest-assured/src/main/java/org/openapitools/client/api/FakeApi.java:[637,37] variable _FORM is already defined in class org.openapitools.client.api.FakeApi.TestEndpointParametersOper
[ERROR] samples/client/petstore/java/rest-assured/src/main/java/org/openapitools/client/api/FakeApi.java:[648,37] variable _FORM is already defined in class org.openapitools.client.api.FakeApi.TestEndpointParametersOper
[ERROR] samples/client/petstore/java/rest-assured/src/main/java/org/openapitools/client/api/FakeApi.java:[659,37] variable _FORM is already defined in class org.openapitools.client.api.FakeApi.TestEndpointParametersOper
[ERROR] samples/client/petstore/java/rest-assured/src/main/java/org/openapitools/client/api/FakeApi.java:[670,37] variable _FORM is already defined in class org.openapitools.client.api.FakeApi.TestEndpointParametersOper
[ERROR] samples/client/petstore/java/rest-assured/src/main/java/org/openapitools/client/api/FakeApi.java:[681,37] variable _FORM is already defined in class org.openapitools.client.api.FakeApi.TestEndpointParametersOper
[ERROR] samples/client/petstore/java/rest-assured/src/main/java/org/openapitools/client/api/FakeApi.java:[692,37] variable _FORM is already defined in class org.openapitools.client.api.FakeApi.TestEndpointParametersOper
[ERROR] samples/client/petstore/java/rest-assured/src/main/java/org/openapitools/client/api/FakeApi.java:[703,37] variable _FORM is already defined in class org.openapitools.client.api.FakeApi.TestEndpointParametersOper
[ERROR] samples/client/petstore/java/rest-assured/src/main/java/org/openapitools/client/api/FakeApi.java:[714,37] variable _FORM is already defined in class org.openapitools.client.api.FakeApi.TestEndpointParametersOper
[ERROR] samples/client/petstore/java/rest-assured/src/main/java/org/openapitools/client/api/FakeApi.java:[725,37] variable _FORM is already defined in class org.openapitools.client.api.FakeApi.TestEndpointParametersOper
[ERROR] samples/client/petstore/java/rest-assured/src/main/java/org/openapitools/client/api/FakeApi.java:[736,37] variable _FORM is already defined in class org.openapitools.client.api.FakeApi.TestEndpointParametersOper
[ERROR] samples/client/petstore/java/rest-assured/src/main/java/org/openapitools/client/api/FakeApi.java:[747,37] variable _FORM is already defined in class org.openapitools.client.api.FakeApi.TestEndpointParametersOper
[ERROR] samples/client/petstore/java/rest-assured/src/main/java/org/openapitools/client/api/FakeApi.java:[758,37] variable _FORM is already defined in class org.openapitools.client.api.FakeApi.TestEndpointParametersOper
[ERROR] samples/client/petstore/java/rest-assured/src/main/java/org/openapitools/client/api/FakeApi.java:[856,36] variable _HEADER is already defined in class org.openapitools.client.api.FakeApi.TestEnumParametersOper
[ERROR] samples/client/petstore/java/rest-assured/src/main/java/org/openapitools/client/api/FakeApi.java:[878,36] variable _QUERY is already defined in class org.openapitools.client.api.FakeApi.TestEnumParametersOper
[ERROR] samples/client/petstore/java/rest-assured/src/main/java/org/openapitools/client/api/FakeApi.java:[889,36] variable _QUERY is already defined in class org.openapitools.client.api.FakeApi.TestEnumParametersOper
[ERROR] samples/client/petstore/java/rest-assured/src/main/java/org/openapitools/client/api/FakeApi.java:[900,36] variable _QUERY is already defined in class org.openapitools.client.api.FakeApi.TestEnumParametersOper
[ERROR] samples/client/petstore/java/rest-assured/src/main/java/org/openapitools/client/api/FakeApi.java:[922,37] variable _FORM is already defined in class org.openapitools.client.api.FakeApi.TestEnumParametersOper
[ERROR] samples/client/petstore/java/rest-assured/src/main/java/org/openapitools/client/api/FakeApi.java:[1070,37] variable _FORM is already defined in class org.openapitools.client.api.FakeApi.TestJsonFormDataOper
[ERROR] samples/client/petstore/java/rest-assured/src/main/java/org/openapitools/client/api/UserApi.java:[497,36] variable _QUERY is already defined in class org.openapitools.client.api.UserApi.LoginUser
```

The reset-assured project on `master` is compiling before your change.

Because of the review delay (sorry about that), can you also merge `master` into your branch (or rebase your commits on top of `master` and force push on your branch)?

File: samples/client/petstore/java/rest-assured/src/main/java/org/openapitools/client/api/FakeApi.java [hunk 1]
```diff
@@ -623,146 +623,146 @@ public TestEndpointParametersOper(RequestSpecBuilder reqSpec) {
             return handler.apply(RestAssured.given().spec(reqSpec.build()).expect().spec(respSpec.build()).when().request(POST, REQ_URI));
         }
 
-         public static final String INTEGER_FORM = "integer";
+         public static final String _FORM = "integer";
```
Comment (jmini) [inline samples/client/petstore/java/rest-assured/src/main/java/org/openapitools/client/api/FakeApi.java]: This produces compile errors

**Comment by jmini** [comment]
By the way to test locally all java clients, I am using a pom file that looks like this:

```xml
<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://maven.apache.org/POM/4.0.0"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
    <modelVersion>4.0.0</modelVersion>
    <groupId>tmp</groupId>
    <artifactId>tmp-build-root</artifactId>
    <packaging>pom</packaging>
    <version>1.0.0-SNAPSHOT</version>
    <modules>

<module>samples/client/petstore/java/feign/</module>
<module>samples/client/petstore/java/google-api-client/</module>
<module>samples/client/petstore/java/jersey1/</module>
<module>samples/client/petstore/java/jersey2-java6/</module>
<module>samples/client/petstore/java/jersey2-java8/</module>
<module>samples/client/petstore/java/jersey2/</module>
<module>samples/client/petstore/java/okhttp-gson-parcelableModel/</module>
<module>samples/client/petstore/java/okhttp-gson/</module>
<module>samples/client/petstore/java/rest-assured/</module>
<module>samples/client/petstore/java/resteasy/</module>
<module>samples/client/petstore/java/resttemplate-withXml/</module>
<module>samples/client/petstore/java/resttemplate/</module>
<module>samples/client/petstore/java/retrofit/</module>
<module>samples/client/petstore/java/retrofit2-play24/</module>
<module>samples/client/petstore/java/retrofit2/</module>
<module>samples/client/petstore/java/retrofit2rx/</module>
<module>samples/client/petstore/java/retrofit2rx2/</module>

    </modules>
</project>
```

**Comment by bmordue** [comment]
Thanks @jmini 
I've back out the change to `modules/openapi-generator/src/main/resources/Java/api.mustache`. I don't understand why this was causing compile errors in the rest-assured sample, so I've just backed out that change to resolve the errors. I was able to build all the samples locally.

**Comment by jmini** [comment]
> Shippable — Run 389 status is FAILED. 

I see this in the logs:

```
UNCOMMITTED CHANGES ERROR
There are uncommitted changes in working tree after execution of 'bin/ensure-up-to-date'
HEAD detached from FETCH_HEAD
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   samples/client/petstore/java/jersey2-java6/src/main/java/org/openapitools/client/api/FakeApi.java
	modified:   samples/client/petstore/java/jersey2-java6/src/main/java/org/openapitools/client/api/PetApi.java
	modified:   samples/client/petstore/java/jersey2-java6/src/main/java/org/openapitools/client/api/StoreApi.java
	modified:   samples/client/petstore/java/jersey2-java6/src/main/java/org/openapitools/client/api/UserApi.java
	modified:   samples/client/petstore/java/jersey2-java8/src/main/java/org/openapitools/client/api/FakeApi.java
	modified:   samples/client/petstore/java/jersey2-java8/src/main/java/org/openapitools/client/api/PetApi.java
	modified:   samples/client/petstore/java/jersey2-java8/src/main/java/org/openapitools/client/api/StoreApi.java
	modified:   samples/client/petstore/java/jersey2-java8/src/main/java/org/openapitools/client/api/UserApi.java
	modified:   samples/client/petstore/java/jersey2/src/main/java/org/openapitools/client/api/FakeApi.java
	modified:   samples/client/petstore/java/jersey2/src/main/java/org/openapitools/client/api/PetApi.java
	modified:   samples/client/petstore/java/jersey2/src/main/java/org/openapitools/client/api/StoreApi.java
	modified:   samples/client/petstore/java/jersey2/src/main/java/org/openapitools/client/api/UserApi.java
	modified:   samples/client/petstore/java/resteasy/src/main/java/org/openapitools/client/api/FakeApi.java
	modified:   samples/client/petstore/java/resteasy/src/main/java/org/openapitools/client/api/PetApi.java
	modified:   samples/client/petstore/java/resteasy/src/main/java/org/openapitools/client/api/StoreApi.java
	modified:   samples/client/petstore/java/resteasy/src/main/java/org/openapitools/client/api/UserApi.java
	modified:   samples/client/petstore/java/resttemplate-withXml/src/main/java/org/openapitools/client/api/FakeApi.java
	modified:   samples/client/petstore/java/resttemplate-withXml/src/main/java/org/openapitools/client/api/PetApi.java
	modified:   samples/client/petstore/java/resttemplate-withXml/src/main/java/org/openapitools/client/api/StoreApi.java
	modified:   samples/client/petstore/java/resttemplate-withXml/src/main/java/org/openapitools/client/api/UserApi.java
	modified:   samples/client/petstore/java/resttemplate/src/main/java/org/openapitools/client/api/FakeApi.java
	modified:   samples/client/petstore/java/resttemplate/src/main/java/org/openapitools/client/api/PetApi.java
	modified:   samples/client/petstore/java/resttemplate/src/main/java/org/openapitools/client/api/StoreApi.java
	modified:   samples/client/petstore/java/resttemplate/src/main/java/org/openapitools/client/api/UserApi.java

no changes added to commit (use "git add" and/or "git commit -a")
Please run 'bin/ensure-up-to-date' locally and commit changes
```

@bmordue: what is your workflow when you create the commit "Ran bin/java-petstore-all.sh again". It is not documented, but it should be:

* Run `mvn verify` on the root of the repo
* Run `.bin/java-petstore-all.sh`
* Commit the changes

Then the check in Shippable will be OK.

**Comment by jmini** [comment]
Thank you for the fix, the build is still red:

> Shippable — Run 437 status is FAILED. 

But this is because of 71b5de3 introduced by #104. I hope this will be fixed ASAP in order to be able to verify that your PR is clean.

**Comment by jmini** [comment]
Commit c30afdf7753d8e2cdf2a4c92aee45f18fcd728b4 fixes the Shippable build. Can you please merge `master` into your branch and push the merge commit or rebase on top of `master` and force push?

**Comment by jmini** [comment]
I have merged `master` into this PR again. I hope this will fix the CI.

**Comment by rubms** [comment]
This pull request is causing an error, that originated issue https://github.com/OpenAPITools/openapi-generator/issues/513: now an empty Object is being set as request body on every request (not only POST requests) and by default message converters don't know how to treat empty Objects, throwing the exception `org.springframework.web.client.RestClientException: Could not write request: no suitable HttpMessageConverter found for request type [java.lang.Object] and content type [application/json]`.

Before this pull request, when post objects were null by default, this error did not happen.

**Comment by rubms** [comment]
I just created the https://github.com/OpenAPITools/openapi-generator/pull/605 pull request in order to get this fixed.

**Comment by MarvGilb** [comment]
I'm concerned. Why aren't there tests to cover this? Simple GET/PUT/POST for every library should be mandatory!

**Pull request merged**